### PR TITLE
Amer/encode strings

### PR DESCRIPTION
**Pull request by amerameen**
## Changes
- Added support for string type in Encode function
- Added tests for strings in encode_test
- Added a helper function to help generating test strings 

## Tests:
Can't run only string ones, so run all encoding tests:
```
go test -run TestEncode
```

## Notes
- I've assumed that UTF-8 is the way strings will get encoded into bytes, but we should confirm this with the other implementers. Initially I assumed ASCII: the encoding that was produced was different than what I expected (both the size of the encoding and the encoding itself)

### CODE REVIEW

**Comment by noot** [comment]
tests pass locally but fail on CI, @ansermino @priom ? 

also pls fix merge conflicts @amerameen

File: codec/encode_test.go [hunk 1]
```diff
@@ -36,13 +37,53 @@ var encodeTests = []encodeTest{
 	{val: false, output: []byte{0x00}},
 }
 
+func setUpStringTests() {
+	// Test strings for various values of n, mode. Also test strings with special characters
+	// TODO: Confirm the UTF-8 is the standard that is being used cross-client
+	// TODO: Flag to omit long running tests for the CI
+	testString1 := "We love you! We believe in open source as wonderful form of giving." 	// n = 67
+	testString2 := strings.Repeat("We need a longer string to test with. Let's multiple this several times.", 230) 		//n = 72 * 230 = 16560
+	testString3 := strings.Repeat("We need a longer string to test with. Let's multiple this several times.", 14913081) 	//n = 72 * 14 913 081 = 1 073 741 832 (> 2^30 = 1 073 741 824)
```
Comment (noot) [inline codec/encode_test.go]: remove comments pls

File: codec/encode_test.go [hunk 2]
```diff
@@ -36,13 +37,53 @@ var encodeTests = []encodeTest{
...
+	testString4 := "Let's test some special ASCII characters: ~  · © ÿ" 	// n = 55 (UTF-8 encoding versus n = 51 with ASCII encoding)
+
+	testStrings := []encodeTest{
+		{val: string("a"),
+			output: []byte{0x04, 0x61}},
+		{val: string("go-pre"),													// n =  6 = 0b110, mode = 0 = 0b00
+			output: append([]byte{0x18}, string("go-pre")...)}, 			 	// n|mode = 0b11000 = 0x18, Enc = n | mode | 0x("go-pre")
```
Comment (noot) [inline codec/encode_test.go]: same for these

File: codec/encode_test.go [hunk 3]
```diff
@@ -36,13 +37,53 @@ var encodeTests = []encodeTest{
...
+	for _, testString := range testStrings {
+		encodeTests = append(encodeTests, testString)
+	}
+}
+
 func TestEncode(t *testing.T) {
+
+	setUpStringTests()
+
+	//Run all tests
```
Comment (noot) [inline codec/encode_test.go]: remove comment pls

File: codec/encode_test.go [hunk 4]
```diff
@@ -31,7 +32,7 @@ var encodeTests = []encodeTest{
 	{val: []byte{0x01, 0x01}, output: []byte{0x08, 0x01, 0x01}, bytesEncoded: 3},
 	{val: []byte{0x01, 0x01}, output: []byte{0x08, 0x01, 0x01}, bytesEncoded: 3},
 	{val: byteArray(64), output: append([]byte{0x01, 0x01}, byteArray(64)...), bytesEncoded: 66},
-	{val: byteArray(16384), output: append([]byte{0x02, 0x00, 0x01, 0x00}, byteArray(16384)...), bytesEncoded: 16388},
+	//{val: byteArray(16384), output: append([]byte{0x02, 0x00, 0x01, 0x00}, byteArray(16384)...), bytesEncoded: 16388},
```
Comment (ansermino) [inline codec/encode_test.go]: Why have you commented this out? test isn't new

File: codec/encode_test.go [hunk 5]
```diff
@@ -74,16 +75,52 @@ var encodeTests = []encodeTest{
 	{val: []*big.Int{big.NewInt(0), big.NewInt(1)}, output: []byte{0x08, 0x00, 0x04}, bytesEncoded: 3},
 }
 
+func setUpStringTests() {
+	// Test strings for various values of n & mode. Also test strings with special characters
+	testString1 := "We love you! We believe in open source as wonderful form of giving." 	// n = 67
+	testString2 := strings.Repeat("We need a longer string to test with. Let's multiple this several times.", 230) 		// n = 72 * 230 = 16560
+	//testString3 := strings.Repeat("We need a longer string to test with. Let's multiple this several times.", 14913081) 	// n = 72 * 14913081 = 1073741832 (> 2^30 = 1073741824)
```
Comment (ansermino) [inline codec/encode_test.go]: Please add comment noting why this commented out (see https://github.com/ChainSafeSystems/go-pre/blob/a89d21793d786cbafec21dd3193ef0870df30945/codec/decode_test.go#L85)

**Comment by priom** [comment]
@amerameen👍🏻

File: codec/encode_test.go [hunk 6]
```diff
@@ -36,13 +37,53 @@ var encodeTests = []encodeTest{
 	{val: false, output: []byte{0x00}},
 }
 
+func setUpStringTests() {
+	// Test strings for various values of n, mode. Also test strings with special characters
+	// TODO: Confirm the UTF-8 is the standard that is being used cross-client
+	// TODO: Flag to omit long running tests for the CI
```
[the remark this answers is not in the document]
Comment (amerameen) [reply codec/encode_test.go]: Done #60

File: codec/encode_test.go (hunk 4 above)
Comment (amerameen) [reply codec/encode_test.go]: yes, this was my mistake! thanks for catching

**Comment by noot** [comment]
@ansermino can this be merged?

**Pull request merged**